app/services/mentee_application_transition_service.rb

```diff
@@ -0,0 +1,72 @@
+module MenteeApplicationTransitionService
+  class InvalidTransitionError < StandardError; end
+  extend self
+
+  STATUS_TRANSITION_MAPPING = {
+    application_received: {
+      valid_transitions: %i[promote reject],
+      promote_transition: :coding_challenge_sent
+    },
+    coding_challenge_sent: {
+      valid_transitions: [:promote],
+      promote_transition: :coding_challenge_received
+    },
+    coding_challenge_received: {
+      valid_transitions: %i[promote reject],
+      promote_transition: :coding_challenge_approved
+    },
+    coding_challenge_approved: {
+      valid_transitions: [:promote],
+      promote_transition: :phone_screen_scheduled
+    },
+    phone_screen_scheduled: {
+      valid_transitions: [:promote],
+      promote_transition: :phone_screen_completed
+    },
+    phone_screen_completed: {
+      valid_transitions: %i[promote reject],
+      promote_transition: :accepted
+    },
+    accepted: {
+      valid_transitions: []
+    },
+    rejected: {
+      valid_transitions: []
+    }
+  }.freeze
+
+  def promote!(application:, user:, note: nil)
+    status = application.current_status.to_sym
+    raise InvalidTransitionError unless STATUS_TRANSITION_MAPPING[status][:valid_transitions].include? :promote
+    promote_transition = STATUS_TRANSITION_MAPPING[status][:promote_transition]
+    application.mentee_application_states.create!(status: promote_transition, status_changed: user, note:)
+    invoke_side_effects(application, promote_transition, user)
+    true
+  end
+
+  def reject!(application:, user:, note: nil)
+    status = application.current_status.to_sym
+    raise InvalidTransitionError unless STATUS_TRANSITION_MAPPING[status][:valid_transitions].include? :reject
+    promote_transition = :rejected
+    application.mentee_application_states.create!(status: promote_transition, status_changed: user, note:)
+    invoke_side_effects(application, promote_transition, user)
+    true
+  end
```

Comment:
A more open question of do we want to keep this stateless like we have now or allow for `application` and `user` to be passed in as instances (and i'll call it semi stateless) but give access to nice instance variables? I am fine either way (I do have a preference but not a big deal)

Reply:
What do you consider the larger benefits being of instantiating the service with application and reviewer? Though I considered instantiating, it seemed more reasonable to reach for a PORO singleton type thing and have the service act as I/O as possible. 

tl;dr My preference is stateless -- I think the state that this service should have is only the state machine transitions, but would be open to better understanding why you might want to pass in the application and user.

Reply:
> 

oh totally fine with that. Was more curious as to your thoughts. I am totally fine with this. I sometimes prefer instance variables so we can use them in other methods more easily (typically I call it like stateless instantiation, because the service should only be run once).